akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -426,104 +424,273 @@ object StreamLayout {
+          case Both(s)                      => try s.onError(ex) catch { case NonFatal(_) => } finally set(Inert)
+          case _                            => // spec violation or cancellation race, but nothing we can do
+        }
+      rec()
+      throw ex // must throw NPE, rule 2:13
+    } else {
+      @tailrec def rec(): Unit =
+        get() match {
+          case Both(s) =>
+            try s.onNext(t)
+            catch {
+              case NonFatal(e) =>
+                set(Inert)
+                throw new IllegalStateException("Subscriber threw exception, this is in violation of rule 2:13", e)
+            }
+          case s: Subscriber[_] => // spec violation
+            val ex = new IllegalStateException(noDemand)
+            getAndSet(Inert) match {
+              case Inert => // nothing to be done
+              case _     => ErrorPublisher(ex, "failed-VirtualProcessor").subscribe(s)
+            }
+            throw ex
+          case Inert | _: Publisher[_] => // nothing to be done
+          case other =>
+            val pub = ErrorPublisher(new IllegalStateException(noDemand), "failed-VirtualPublisher")
+            if (!compareAndSet(other, pub)) rec()
+            else throw pub.t
+        }
+      rec()
     }
-    @tailrec
+
+  private def noDemand = "spec violation: onNext was signaled from upstream without demand"
+
+  private class WrappedSubscription(real: Subscription) extends Subscription {
     override def request(n: Long): Unit = {
-      val current = get
-      if (current < 0) s.request(n)
-      else if (compareAndSet(current, current + n)) ()
-      else request(n)
+      if (n < 1) {
+        tryCancel(real)
+        getAndSet(Inert) match {
+          case Both(s) => rejectDueToNonPositiveDemand(s)
+          case Inert   => // another failure has won the race
+          case _       => // this cannot possibly happen, but signaling errors is impossible at this point
+        }
+      } else real.request(n)
+    }
+    override def cancel(): Unit = {
+      set(Inert)
+      real.cancel()
     }
-    def closeLatch(): Unit = {
-      val requested = getAndSet(-1)
-      if (requested > 0) s.request(requested)
+  }
+}
+
+/**
+ * INTERNAL API
+ *
+ * The implementation of `Sink.asPublisher` needs to offer a `Publisher` that
+ * defers to the upstream that is connected during materialization. This would
+ * be trivial if it were not for materialized value computations that may even
+ * spawn the code that does `pub.subscribe(sub)` in a Future, running concurrently
+ * with the actual materialization. Therefore we implement a minimial shell here
+ * that plugs the downstream and the upstream together as soon as both are known.
+ * Using a VirtualProcessor would technically also work, but it would defeat the
+ * purpose of subscription timeouts—the subscription would always already be
+ * established from the Actor’s perspective, regardless of whether a downstream
+ * will ever be connected.
```

Comment:
Yeah. The other option would be to move the special casing to the Actor itself, which would recognize that a VirtualProcessor is attached and would modify its behavior accordingly. But that is rather ugly, so I guess this is what we have then.